generator/internal/parser/httprule/http_rule_parser_test.go

```diff
@@ -0,0 +1,136 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httprule
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"testing"
+)
+
+func TestProtobuf_Parse(t *testing.T) {
+	tests := []struct {
+		path        string
+		want        *PathTemplate
+		explanation string
+	}{
+		{"/foo", template(literal("foo")), ""},
+		{"/foo/bar", template(literal("foo"), literal("bar")), ""},
+		{"/v1/*/foo", template(literal("v1"), match(), literal("foo")), ""},
+		{"/v1/**/foo", template(literal("v1"), matchR(), literal("foo")), ""},
+		{"/foo:bar", templateV(segments(literal("foo")), verb("bar")), ""},
+		{"/foo/{bar}", template(literal("foo"), varr("bar")), ""},
+		{"/foo/{bar.baz}", template(literal("foo"), varr("bar", "baz")), ""},
+		{"/foo/{bar=baz}", template(literal("foo"), varrs(ids("bar"), segments(literal("baz")))), ""},
+		{"/foo/{bar=*}", template(literal("foo"), varrs(ids("bar"), segments(match()))), ""},
+		{"/foo/{bar=*}/baz", template(literal("foo"), varrs(ids("bar"), segments(match())), literal("baz")), ""},
+		{"/foo/{bar=*}/baz:qux", templateV(segments(literal("foo"), varrs(ids("bar"), segments(match())), literal("baz")), verb("qux")), ""},
+		{"foo", nil, "path must start with slash"},
+		{"/", nil, "path cannot end with slash"},
+		{"/foo/", nil, "path cannot end with slash"},
+
+		// the following test is failing because * is a sub-delimiter, which is allowed in the LITERAL segment
+		//{"/foo/***/bar", nil, "wildcard literal cannot exceed two *"},
+
+		{"/%0f", template(literal("%0f")), ""},
+		{"/%0z", nil, "bad percent encoding"},
+		{"/foo//bar", nil, "segment is too short"},
+		{"/foo/:", nil, "verb is too short"},
+		{"/foo/{}/bar", nil, "var too short"},
+		{"/foo/{a.}/bar", nil, "var identifier too short"},
+		{"/foo/{.a}/bar", nil, "var identifier too short"},
+		{"/foo/{a=}/bar", nil, "var value too short"},
+		{"/foo/{9bar}", nil, "var identifier has bad first character"},
+		{"/foo/{bar9}", template(literal("foo"), varr("bar9")), ""},
+		{"/foo/{b&r}", nil, "var identifier has bad character"},
+		{"/foo/:bar", nil, "verb cannot come after slash"},
+		{"/foo:bar/baz", nil, "verb must be the last segment"},
+		{":foo", nil, "verb cannot be the first segment"},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.path, func(t *testing.T) {
+			got, err := Parse(tc.path)
+			if tc.want != nil {
+				if err != nil {
+					t.Fatalf("expected no error, got: %v", err)
+				}
+				if got == nil {
+					t.Fatalf("expected path template for %s, got nil", tc.path)
+				}
+				if diff := cmp.Diff(tc.want, got, cmpopts.EquateEmpty()); diff != "" {
+					t.Fatalf("failed parsing path [%s] (-want, +got):\n%s", tc.path, diff)
+				}
+			} else {
+				if err == nil {
+					t.Fatalf("Parse(%s) succeeded, want error: %s", tc.path, tc.explanation)
+				}
+			}
+		})
+	}
+}
+
```

Comment:
I could get some feedback on the functions below, and whether we think they help or hinder readability (as well as suggestions for better names 🙂).

Reply:
(1) they are in a test file, and only used there, the names are not *that* important.
(2) I wonder if we could write these as:

```golang
func expectTemplate() *PathTemplate {
  return &PathTemplate{}
}

func (p *PathTemplate) withLiteral(string v) *PathTemplate {
  p.Segments = append(p.Segments, /* may need some magic */ v)
  return p
}

func (p *PathTemplate) withVerb(string v) *PathTemplate) {
  p.Verb = (Literal*)&v;
  return p
}

// And so on.
```

Reply:
Nice!